### PR TITLE
fix: added support for getting/setting a Date object

### PR DESCRIPTION
**Pull request by Sadeka**
Hello,

I have updated the code to add support for getting/setting a Date object (works for IndexedDB). I have also added a unit test for for it. Could you please review my changes?

Thanks & Regards,
Sadeka

### CODE REVIEW

File: src/angular-localForage.js [hunk 1]
```diff
@@ -143,7 +143,11 @@
           var copy;
           if (angular.isArray(value)) {
             return value.map(stripMeta);
-          } else if (angular.isObject(value) && !(value instanceof Blob)) {
+          } 
+          else if(angular.isObject(value) && (value instanceof Date)) {
+            return value;
+          }
+          else if (angular.isObject(value) && !(value instanceof Blob)) {
```
Comment (scotttrinh) [inline src/angular-localForage.js]: Style nit: Will you please put the `else`s and `else if`s on the same line as the closing brace of the `if` and make sure there is a space between `else`/`else if` and the opening `(` of the conditional statement? Like:

```js
} else if (false === true) {
```

File: src/angular-localForage.js [hunk 2]
```diff
@@ -171,7 +175,7 @@
           args = arguments,
           self = this,
           promise;
-
+
```
Comment (scotttrinh) [inline src/angular-localForage.js]: Extra whitespace here.

File: src/angular-localForage.js [hunk 3]
```diff
@@ -198,11 +202,13 @@
               deferred.reject(res);
             }
           });
-        } else {
+        } 
+        else {
```
Comment (scotttrinh) [inline src/angular-localForage.js]: `} else {`

File: src/angular-localForage.js [hunk 4]
```diff
@@ -198,11 +202,13 @@
...
           promise = self._localforage.getItem(self.prefix() + key).then(function(item) {
             if (rejectOnNull && item === null) {
               deferred.reject(item);
-            } else {
+            } 
+            else {
```
Comment (scotttrinh) [inline src/angular-localForage.js]: `} else {`

File: tests/angular-localForage.js [hunk 5]
```diff
@@ -183,20 +183,41 @@ describe('Module: LocalForageModule', function() {
 
   it('setItem and getItem should work', function(done) {
     var interval = triggerDigests();
+    const timestamp = new Date();
```
Comment (scotttrinh) [inline tests/angular-localForage.js]: No ES2015 in this codebase for now. I plan on doing a good sprint on this where at least the tests get updated, but until we change it all, let's stick with `var`.

File: tests/angular-localForage.js [hunk 6]
```diff
@@ -183,20 +183,41 @@ describe('Module: LocalForageModule', function() {
...
+        expect(data).toEqual({ name: 'Scott Trinh', date: timestamp });
+        done();
+      }, done);
+
+    }, done);
+  });
+
+  it('setItem and getItem should work for a Date object', function(done) {
+    var interval = triggerDigests();
+    const timestamp = new Date();
```
Comment (scotttrinh) [inline tests/angular-localForage.js]: `const` -> `var`

File: tests/angular-localForage.js [hunk 7]
```diff
@@ -183,20 +183,41 @@ describe('Module: LocalForageModule', function() {
...
+
+    spyOn($localForage._localforage, 'setItem').and.callThrough();
+
+    $localForage.setItem('myDateObject', timestamp).then(function(data) {
+      expect(data).toEqual(timestamp);
+      expect($localForage._localforage.setItem.calls.mostRecent().args[1]).toEqual( timestamp );
```
Comment (scotttrinh) [inline tests/angular-localForage.js]: No need to pad `timestamp` with extra spaces. `(timestamp)` is perfectly fine!

**Comment by scotttrinh** [comment]
Actually, @Sadeka , I think there is a better way to accomplish this. Instead of special-casing all non Primitive Objects, we should just change the check to see if the `.constructor` is `Object`. Keep all of the tests you wrote, but would you remove the special case for `Date` and change L#150 to:

```js
} else if (angular.isObject(value) && value.constructor === Object) {
```

That will fail for `Blob`s and `Date`s and `moment`s and anything else that uses a special constructor.

Let me know if that makes sense or if you see a logical flaw in that!

**Comment by Sadeka** [comment]
Thanks @scotttrinh for your valuable feedback. I have updated the code accordingly. Submitting a new pull request with a fresh single commit.

**Pull request closed without merging**